presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,424 @@
+                result = new JoinEnumerationResult(costProvider.getCumulativeCost(resultNode), Optional.of(resultNode));
+            }
+
+            return result;
+        }
+
+        private JoinEnumerationResult getJoinSource(List<PlanNode> nodes, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                ImmutableList.Builder<Expression> predicates = ImmutableList.builder();
+                predicates.addAll(allInference.generateEqualitiesPartitionedBy(outputSymbols::contains).getScopeEqualities());
+                stream(EqualityInference.nonInferrableConjuncts(allFilter))
+                        .map(conjuct -> allInference.rewriteExpression(conjuct, outputSymbols::contains))
+                        .filter(Objects::nonNull)
+                        .forEach(predicates::add);
+                Expression filter = combineConjuncts(predicates.build());
+                if (!(TRUE_LITERAL).equals(filter)) {
+                    planNode = new FilterNode(idAllocator.getNextId(), planNode, filter);
+                }
+                return new JoinEnumerationResult(costProvider.getCumulativeCost(planNode), Optional.of(planNode));
+            }
+            return chooseJoinOrder(nodes, outputSymbols);
+        }
+
+        private static boolean isJoinEqualityCondition(Expression expression)
+        {
+            return expression instanceof ComparisonExpression
+                    && ((ComparisonExpression) expression).getType() == EQUAL
+                    && ((ComparisonExpression) expression).getLeft() instanceof SymbolReference
+                    && ((ComparisonExpression) expression).getRight() instanceof SymbolReference;
+        }
+
+        private static JoinNode.EquiJoinClause toEquiJoinClause(ComparisonExpression equality, Set<Symbol> leftSymbols)
+        {
+            Symbol leftSymbol = Symbol.from(equality.getLeft());
+            Symbol rightSymbol = Symbol.from(equality.getRight());
+            JoinNode.EquiJoinClause equiJoinClause = new JoinNode.EquiJoinClause(leftSymbol, rightSymbol);
+            return leftSymbols.contains(leftSymbol) ? equiJoinClause : equiJoinClause.flip();
+        }
+
+        private JoinEnumerationResult setJoinNodeProperties(JoinNode joinNode)
+        {
+            // TODO avoid stat (but not cost) recalculation for all considered (distribution,flip) pairs, since resulting relation is the same in all case
+
+            List<JoinEnumerationResult> possibleJoinNodes = new ArrayList<>();
+            FeaturesConfig.JoinDistributionType joinDistributionType = getJoinDistributionType(session);
+
+            if (isAtMostScalar(joinNode.getRight(), lookup)) {
+                JoinNode node = joinNode.withDistributionType(REPLICATED);
+                return new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node));
```

Comment:
`new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node))` code is repeated multiple times; perhaps, extract it to a helper method

The helper method could also check for unknown or infinite cost and return `UNKNOWN_COST_RESULT` or `INFINITE_COST_RESULT`. This way, `JoinEnumerationResult` could either have a plan with known cost or not have a plan. This way, checks like this

```
           if (leftResult.cost.hasUnknownComponents()) {
                return UNKNOWN_COST_RESULT;
            }
            if (leftResult.cost.equals(INFINITE_COST)) {
                return INFINITE_COST_RESULT;
            }
```

can be simplified to

```
           if (!leftResult.getPlanNode().isPresent()) {
                return UNKNOWN_COST_RESULT;
            }
```